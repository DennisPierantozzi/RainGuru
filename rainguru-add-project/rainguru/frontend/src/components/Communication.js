import React from "react";
import Popup from "react-popup";
import Map from "./Map/Map";

export default class Communication {
    static useLatestData = true;
    static observed = false;
    static dataCollected = false;
    static imageUrls = [];
    static latestTimestamp = 0;
    static requestTimestamp = [];
    static dataTimestamp = [];

    // getters
    /**
     * Get the boolean representing if we use the latest data.
     */
    static getUseLatestData() {
        return this.useLatestData;
    }

    /**
     * Get the boolean representing if we show observed or predicted data.
     */
    static getObserved() {
        return this.observed;
    }

    /**
     * Get the boolean representing if all the data has been loaded.
     */
    static getDataCollected() {
        return this.dataCollected;
    }

    /**
     * Get the urls for the images with the predicted precipitation.
     */
    static getImageUrls() {
        return this.imageUrls;
    }

    /**
     * Get the timestamp of the first prediction of the latest data.
     */
    static getLatestTimestamp() {
        return this.latestTimestamp;
    }

    /**
     * Get the timestamp that is used when making requests.
     */
    static getRequestTimestamp() {
        return this.requestTimestamp;
    }

    /**
     * Get the timestamp of the first prediction of the data in use.
     */
    static getDataTimestamp() {
        return this.dataTimestamp;
    }

    // setters
    /**
     * Set the boolean representing to if we use the latest data.
     */
    static setUseLatestData(latest) {
        this.useLatestData = latest;
    }

    /**
     * Set the boolean representing if we show observed or predicted data.
     */
    static setObserved(observed) {
        this.observed = observed;
    }

    /**
     * Set the timestamp that the requests will use
     */
    static setRequestTimestamp(timestamp) {
        this.requestTimestamp = timestamp;
    }


    /**
     * Fetches new data from the back-end.
     */
    static async fetchUrls() {
        let url = "api/fetch?observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }
        // make request to retrieve predictions
        await fetch(url)
          .then(response => {
            if (response.status > 400) {
              throw new Error();
            }
            return response.json();
          })
          .then(responseJson => {
            // store predictions and create images for them
            this.imageUrls = responseJson.urls;
            this.dataTimestamp = new Date(responseJson.timestamp * 1000);
            if(responseJson.exception_active) {
                Popup.alert("Message from server:\n" + responseJson.exception_message, "Warning: Model might be using old data.");
                console.log(responseJson.exception_message);
            }
            Map.preloadedImages();
            this.dataCollected = true;
          });
    }

    /**
     * Fetches the timestamp of the first prediction from the latest data.
     *
     * @returns {Promise<boolean>} representing if there is new data available.
     */
    static async checkNewData() {
        let dataUpdated = false;
        // make a request for the latest data first prediction timestamp
        await fetch("api/check")
            .then(response => {
                if (response.status > 400) {
                  throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // check if there is new data via the timestamp received
                const newTimestamp = responseJson.timestamp;
                if (newTimestamp !== this.latestTimestamp) {
                    this.latestTimestamp = newTimestamp;
                    dataUpdated = true;
                }
            });
        return dataUpdated;
    }

    /**
     * Fetches the data about a selected location for a selected 100 minutes.
     *
     * @param x the x coordinate of the 480x480 images
     * @param y the y coordinate of the 480x480 images
     * @returns {Promise<any>} the precipitation of the location for a selected 100 minutes.
     */
    static async fetchPrecipitation(x, y) {
        let url = "api/precipitation?x=" + x + "&y=" + y + "&observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }

        // make a request for the precipitation information of the give location
        return await fetch(url)
            .then(response => {
                if (response.status > 400) {
                    throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // return the fetched precipitation
                return responseJson.precipitation;
            })
    }

    /**
     * Fetch the times of what data is all available
     *
     * @returns {Promise<{available: *, predictedTimes: *, observedTimes: *}>} the availability of data times
     */
    static async fetchDataAvailability() {
        // make a request to get the availability times
        return await fetch("api/available")
            .then(response => {
                if (response.status > 400) {
                    throw new Error()
                }
                return response.json();
            })
            .then(responseJson => {
                // return a boolean representing if data is being stored and the available data times
                return {
                    available: responseJson.store_data,
                    predictedTimes: responseJson.predictions_stored,
                    observedTimes: responseJson.observations_stored
                };
            })
    }
}