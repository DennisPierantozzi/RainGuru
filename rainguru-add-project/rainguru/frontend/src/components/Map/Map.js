import React from "react";
import Popup from "react-popup";
import L from "leaflet";
import {GeoSearchControl, OpenStreetMapProvider} from "leaflet-geosearch";

import Slider from "../Information/Slider";
import Communication from "../Communication";

export default class Map {

    // static variables that contain everything for the map component
    static map;
    static heatLayers = [];
    static nextLayers = [];
    static currentLayer = new L.LayerGroup();
    static marker;
    static loadedImages = 0;
    static dataURLS = [];
    static minMapBounds = L.latLng(49.18, -4.35);
    static maxMapBounds = L.latLng(55.72, 15.18);
    static minMarkerBounds = L.latLng(50.21, 1.90);
    static maxMarkerBounds = L.latLng(54.71, 9.30);


    /**
     * Preload and cache the images
     */
    static preloadedImages() {
        // get the image urls
        const imageUrls = Communication.getImageUrls();

        // for each url preload the image so it doesn't flicker
        imageUrls.forEach((image) => {
            const newImage = new Image();
            newImage.onload = function () {
                Map.loadedImages += 1;
            }
            newImage.src = image;
            window[image] = newImage;
        });
    }

    /**
     * Generates a list of all generated precipitation images for the coming 100 minutes.
     */
    static setHeatLayers() {
        // heat layer code
        const imageUrls = Communication.getImageUrls();
        const imageBounds = [[54.71, 1.90], [50.21, 9.30]];
        this.heatLayers = [];

        // adds the regular image layers
        for (let i = 0; i < 20; i++) {
            const newUrl = imageUrls[i];
            this.heatLayers.push(L.imageOverlay(newUrl, imageBounds, {opacity: 0.4}));
        }

        // adds the image layers that are used to preload the next layer
        for (let i = 1; i < 20; i++) {
            const newUrl = imageUrls[i];
            this.nextLayers.push(L.imageOverlay(newUrl, imageBounds, {opacity: 0}));
        }
        const zeroUrl = imageUrls[0];
        this.nextLayers.push(L.imageOverlay(zeroUrl, imageBounds, {opacity: 0}));
    }

    /**
     * Shows the requested precipitation image on the map.
     * @param layerIndex the index of the requested precipitation image in the 'heatLayers' list.
     */
    static activateHeatLayer(layerIndex) {
        // remove the layers currently showing
        // then add the current layer
        // and add the next layer with opacity 0 to fix the flickering on safari and firefox
        this.currentLayer.clearLayers();
        this.currentLayer.addLayer(this.heatLayers[layerIndex]);
        this.currentLayer.addLayer(this.nextLayers[layerIndex]);
    }

    /**
     * Renders the map and handles the event of clicking on the map and putting a marker at that location.
     */
    static render() {
        // base layer code
        const baseLayer = L.tileLayer('https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/{z}/{y}/{x}', {
            attribution: '&copy; <a href="https://www.esri.com/en-us/home">Esri</a>'
        });

        // set up the heat layers
        Map.setHeatLayers();

        // map code
        let map = L.map("map", {
            center: new L.LatLng(52.45, 5.415),
            zoom: 7,
            minZoom: 7,
            maxBounds: L.latLngBounds(Map.minMapBounds, Map.maxMapBounds),
            maxBoundsViscosity: 0.9,
        });

        // add layers to the map and the scale
        baseLayer.addTo(map);
        this.currentLayer.addTo(map);
        L.control.scale({imperial: false}).addTo(map);

        // load the first rain image and center the map to the Netherlands
        this.activateHeatLayer(0);

        // store the map
        Map.map = map;

        // add an error message for the users current location
        map.on('locationerror', function (e) {
            Popup.alert("User denied access to their location", "Permission denied");
        });

        // add location finding handling to the map
        map.on('locationfound', function (e) {
            Map.markerHandling(e, false, true);
        });

        // add clicking to place a marker to the map
        map.on('click', function (e) {
            Map.markerHandling(e, false, false);
        });

        // add the search functionality to the map
        Map.addMapSearch();

        // When a location has been found this event will be triggered
        map.on('geosearch/showlocation', function (e) {
            Map.markerHandling(e, true, false);
        });
    }

    /**
     * Add the search functionality to the map.
     */
    static addMapSearch() {
        // Adds searching functionality to the map
        const search = new GeoSearchControl({
            provider: new OpenStreetMapProvider(),
            position: 'topleft',
            animateZoom: false,
            retainZoomLevel: true,
            marker: {
                icon: new L.Icon.Default(),
                draggable: false,
            },
            searchLabel: 'Search',
            updateMap: true,
        });
        Map.map.addControl(search);

        // Move the search bar to the top bar
        document.getElementById('locationField').appendChild(
            document.querySelector(".geosearch")
        );
    }

    // method for the map.on events
    /**
     * Handles the placement of the map markers in various situations.
     *
     * @param e the event.
     * @param search a boolean representing if the method was called for a geosearch event.
     * @param found a boolean representing if the method was called for when a location was found (otherwise it is on click).
     */
    static markerHandling(e, search, found) {
        // the location of where we want to try to place a marker
        let latitude;
        let longitude;
        if (search) {
            latitude = e.location.y;
            longitude = e.location.x;
        } else {
            latitude = e.latlng.lat;
            longitude = e.latlng.lng;
        }

        // check if the location is inside the bounds
        if (latitude > Map.maxMarkerBounds.lat || latitude < Map.minMarkerBounds.lat || longitude < Map.minMarkerBounds.lng || longitude > Map.maxMarkerBounds.lng) {
            Popup.alert("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
            if (search) {
                Map.map.setView(Map.marker !== undefined ? Map.marker.getLatLng() : L.latLng(52.45, 5.415), 8);
            }
            return;
        }

        // remove any existing markers
        if (Map.marker !== undefined) {
            Map.map.removeLayer(Map.marker);
        }

        // add the marker
        Map.marker = L.marker([latitude, longitude]).addTo(Map.map);
        // go to the new marker, this is handled differently depending on what event called the method
        if (search) {
            Map.map.setZoom(8);
        } else {
            Map.map.flyTo(L.latLng(latitude, longitude), found ? Math.max(8, Map.map.getZoom()) : Map.map.getZoom());
        }
        Slider.showPrecipitationData(latitude, longitude);
    }
}
