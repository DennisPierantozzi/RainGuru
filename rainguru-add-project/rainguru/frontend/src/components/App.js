import React, {Component} from "react";
import {createRoot} from "react-dom/client";
import LoadingScreen from "./LoadingScreen";
import Communication from "./Communication";
import TopBar from "./TopBar/TopBar";
import PastDataSelector from "./TopBar/PastDataSelector";
import Map from "./Map/Map";
import Information from "./Information/Information";
import AnimationBar from "./Information/AnimationBar";
import Slider from "./Information/Slider";
import PopupContainer from "./PopupContainer/PopupContainer";
import SideBar from "./TopBar/SideBar";
import InfoMenu from "./TopBar/InfoMenu";


export default class App extends Component {
    /**
     * Sets the app in the starting state.
     * @param props contains optional properties of the app.
     */
    constructor(props) {
        super(props);

        // state variables
        this.state = {
            loadingData: false,
            animationBarUpdateProp: 0,
            pastDataSelectorUpdateProp: 0,
            sidebar: false,
            compare: false,
            sidebarInfo: false,
        }

        // bind this to update the state when this function is called from the past data selector
        this.displayData = this.displayData.bind(this);
        this.compareData = this.compareData.bind(this);
        this.displayComparedData = this.displayComparedData.bind(this);
        this.setShowSidebar = this.setShowSidebar.bind(this);
        this.setShowCompare = this.setShowCompare.bind(this);
        this.setShowSidebarInfo = this.setShowSidebarInfo.bind(this);
    }

    static renderSetup() {
        const waitForSetupDOM = setInterval(function() {
            const popupContainerDiv = document.getElementById("popupContainer");
            if (!!popupContainerDiv) {
                // render the popup container in case we need to display loading errors
                const popupContainerRoot = createRoot(popupContainerDiv);
                popupContainerRoot.render(<PopupContainer/>);

                // retrieve the data from the server
                Communication.checkNewData();
                Communication.fetchUrls();


                // wait for the map to fetch the data
                const waitForDataLoad = setInterval(function () {
                    if (Communication.getDataCollected() && Map.loadedImages === 20) {
                        // end this loop render the components
                        clearInterval(waitForDataLoad);
                        const appDiv = document.getElementById("app");
                        const appRoot = createRoot(appDiv);
                        appRoot.render(<App/>);
                        clearInterval(LoadingScreen.loadingIntervalId);
                    }
                }, 50);

                clearInterval(waitForSetupDOM);
            }
        }, 50);
    }

    setShowSidebar() {
        this.setState({
            sidebarInfo: false,
            sidebar: !this.state.sidebar            
        })
    }

    setShowSidebarInfo() {
        this.setState({ 
            sidebar: false,
            sidebarInfo: !this.state.sidebarInfo
        })
    }

    setShowCompare(compare) {
        this.setState({compare: compare});
    }


    /**
     * Renders the components that the screen is divided into and set the setup to load the map into.
     */
    render() {
        return [
            <div id="map" key="map"></div>,
            <div id="information" key="information"><Information animationBarUpdateProp={this.state.animationBarUpdateProp}/></div>,
            <div id="topBar" key="topBar"><TopBar
                                     setShowSidebar = {this.setShowSidebar} displayComparedData={this.displayComparedData}
                                     loadingData={this.state.loadingData} showCompare={this.state.compare} displayData={this.displayData} 
                                     setShowSidebarInfo={this.setShowSidebarInfo}/></div>,
            <div id="menu" key="menuOverlay"  className={this.state.sidebar ? "menuOverlay" : "hideElement"}>
                <SideBar pastDataSelectorUpdateProp={this.state.pastDataSelectorUpdateProp}
                    loadingData={this.state.loadingData} displayData={this.displayData} 
                    compareData={this.compareData} 
                    displayComparedData={this.displayComparedData} setShowCompare={this.setShowCompare} setShowSidebar={this.setShowSidebar}/>
            </div>, 
            <div id="menu-info" key="menuOverlay-info" className={this.state.sidebarInfo ? "menuOverlay" : "hideElement"}>
                <InfoMenu />
            </div>
        ];
    }

    /**
     * Renders all the parts of the application.
     */
    componentDidMount() {
        // render the map
        // this is done last since it calls on other parts which could cause problems
        // another reason that it is done after the initial render is since it renders itself inside the map div
        Map.render();

        // prevent keyboard from resizing the application
        const viewport = document.querySelector("meta[name=viewport]");
        viewport.setAttribute("content", viewport.content + ", height=" + window.innerHeight);

        try {
            // Reset this on orientation change to prevent resizing issues
            window.screen.orientation.onchange = function(e) {
                const viewport = document.querySelector("meta[name=viewport]");

                let maxSize = Math.max(window.innerWidth, window.innerHeight);
                let minSize = Math.min(window.innerWidth, window.innerHeight);

                // Set the height according to the orientation
                if (e.target.type === "landscape-primary" || e.target.type === "landscape-secondary") {
                    viewport.setAttribute("content", viewport.content + ", height=" + minSize);
                } else {
                    viewport.setAttribute("content", viewport.content + ", height=" + maxSize);
                }
            }
        } catch (error) {
            try {
                // Reset this on orientation change on ios devices
                window.onorientationchange = function(e) {
                    const viewport = document.querySelector("meta[name=viewport]");

                    let maxSize = Math.max(window.innerWidth, window.innerHeight);
                    let minSize = Math.min(window.innerWidth, window.innerHeight);

                    // Set the height according to the orientation
                    if (Math.abs(window.orientation) === 90) {
                        viewport.setAttribute("content", viewport.content + ", height=" + minSize);
                    } else {
                        viewport.setAttribute("content", viewport.content + ", height=" + maxSize);
                    }
                    
                }
            } catch (error) {
                console.error(error);
            }
        }

        // start the auto updating
        this.runAutoUpdate();
    }

    /**
     * Checks for new data and updates all parts of the application that need it when there is new data.
     */
    runAutoUpdate() {
        // check for new available data every 60 seconds
        setInterval(function () {
            Communication.checkNewData()
                .then(dataUpdated => {
                    // if there is new data try to update
                    if (dataUpdated) {
                        // update the time intervals of the past data menu
                        PastDataSelector.updateTimes();
                        displayData(-1, true, false, [], "Latest data");
                        this.setState({pastDataSelectorUpdateProp: this.state.pastDataSelectorUpdateProp + 1});

                        // update the rain layers and animation bar timestamps when using the latest data
                        if (Communication.getUseLatestData()) {
                            this.displayData(1, true, false, []);
                        }
                    }
                });
        }.bind(this), 60000);
    }

    /**
     * Changes the map rain layers and the animation bar timestamps to new data that is requested.
     *
     * @param propChange the change that the animation bar prop should update which indicates how it should update.
     * @param latest a boolean value representing if the latest data should be loaded.
     * @param observed a boolean value representing if observed or predicted data should be loaded.
     * @param timestamp the timestamp from which to load the data in case it isn't the latest.
     */
    displayData(propChange, latest, observed, timestamp, timeString) {
        // only run this code if it isn't already loading data already
        if (!this.state.loadingData) {
            this.setState({loadingData: true, compare: false});
            document.getElementById("currently-showing").innerHTML = timeString;
            // set the communication variables used in the requests
            Communication.setCompare(false);
            Communication.setUseLatestData(latest);
            Communication.setObserved(observed);
            Communication.setRequestTimestamp(timestamp);

            // reset the loaded images count and then fetch the new images
            Map.loadedImages = 0;

            // Set loading data visible
            const loader = document.getElementById("loader");
            document.getElementById("loadingString").innerHTML = "";
            loader.style.display = "flex";

            Communication.fetchUrls()
                .then(() => {
                    // wait for all new images to be preloaded and then render them and update the other parts
                    const waitForNewImages = setInterval(function () {
                        if (Map.loadedImages === 20) {

                            // hide loading data visible
                            loader.style.display = "none";
                            
                            // set the rain images
                            Map.setHeatLayers();
                            
                            Slider.getPrecipitationData(Slider.lastLat, Slider.lastLong);

                            // update the animation bar timestamps and past data time intervals
                            AnimationBar.updateMarks();
                            this.setState({animationBarUpdateProp: this.state.animationBarUpdateProp + propChange});

                            // clear the interval and set the loading to be done
                            this.setState({loadingData: false});
                            clearInterval(waitForNewImages);
                        }
                    }.bind(this), 100);
                });
        }
    }


/**
     * Changes the map rain layers and the animation bar timestamps to new data that is requested
     * Prepare to compare predictions and observations
     *
     * @param propChange the change that the animation bar prop should update which indicates how it should update.
     * @param latest a boolean value representing if the latest data should be loaded.
     * @param observed a boolean value representing if observed or predicted data should be loaded.
     * @param timestamp the timestamp from which to load the data in case it isn't the latest.
     * @param compare boolean value that indicate the start of the compare feature
     */
    compareData(propChange, latest, observed, timestamp, timeString) {
        // only run this code if it isn't already loading data already
        if (!this.state.loadingData) {
            this.setState({loadingData: true});
            // set the communication variables used in the requests
            Communication.setUseLatestData(latest);
            Communication.setObserved(observed);
            Communication.setRequestTimestamp(timestamp);
            Communication.setCompare(true);

            // reset the loaded images count and then fetch the new images
            Map.loadedImages = 0;
            Communication.fetchUrlsToCompare()
                .then(() => {
                    // wait for all new images to be preloaded and then render them and update the other parts
                    const waitForNewImages = setInterval(function () {
                        if (Map.loadedImages === 20) {
                            // display switch
                            this.setState({compare: true});

                            // set the timeString
                            const currentlyShowing = document.getElementById("currently-showing");
                            currentlyShowing.innerHTML = timeString;

                            // set the rain images
                            Map.setHeatLayers();

                            // update the animation bar rain chart and location information
                            if (Slider.lastLat !== undefined) {
                                Slider.getPrecipitationData(Slider.lastLat, Slider.lastLong);
                            }

                            // update the animation bar timestamps and past data time intervals
                            AnimationBar.updateMarks();
                            this.setState({animationBarUpdateProp: this.state.animationBarUpdateProp + propChange});

                            // clear the interval and set the loading to be done
                            this.setState({loadingData: false});
                            clearInterval(waitForNewImages);
                        }
                        else {
                            // error, no intervals for predictions
                            this.setState({loadingData: false});
                            this.setState({compare: false});
                        }
                    }.bind(this), 100);
                });
        }
    }


    /**
     * Changes the map rain layers based on the compare switch
     *
     * @param pred boolean value that indicate the layers to display. 
     *        False for observation, True for predictions
     */
    displayComparedData(pred) {
        // only run this code if it isn't already loading data already
        if (!this.state.loadingData) {
            this.setState({loadingData: true});
            // set the communication variables used in the requests
            Map.loadedImages = 0;
            Map.preloadedImages(pred);
            const waitForNewImages = setInterval(function () {
                if (Map.loadedImages === 20) {
                    // set the rain images
                    Map.setHeatLayers();

                    //graph precipitation switch
                    //Slider.switchPrediction();
                    //Slider.colorLegend(pred);
                    if (Slider.lastLat !== undefined) {
                       Slider.showPredictionData(pred);
                    }

                    // update the animation bar timestamps and past data time intervals
                    AnimationBar.updateMarks();
                    this.setState({animationBarUpdateProp: this.state.animationBarUpdateProp -1});

                    // clear the interval and set the loading to be done
                    this.setState({loadingData: false});
                    clearInterval(waitForNewImages);
                }
            }.bind(this), 100);
        }
    }

}
