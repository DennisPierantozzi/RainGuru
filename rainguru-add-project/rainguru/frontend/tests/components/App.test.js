import React from "react";
import {render, screen} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import App from "../../src/components/App";
import Communication from "../../src/components/Communication";
import Map from "../../src/components/Map/Map";

const imageUrlsUsed = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
        "10", "11", "12", "13", "14", "15", "16", "17", "18", "19"]

beforeEach(() => {
    // set up communication with fake image urls
    Communication.imageUrls = imageUrlsUsed;

    // mock the search functionality since it gives problems when testing
    Map.addMapSearch = jest.fn();

    // set up communication with a timestamp (23:50), so it can't break the tests since it will be requested in most
    // use a correction for timezones so each timezone runs it with the same local time to keep the tests intact
    const timeZoneOffset = ((new Date()).getTimezoneOffset()) * 60;
    Communication.dataTimestamp = new Date(1000 * (1653263400 + timeZoneOffset));
});

describe("App tests", () => {
    it("matches snapshot", () => {
        const app = new App();
        app.componentDidMount = jest.fn(() => {
            Map.render();
        });

        render(<div id="main" data-testid="main">app.render()</div>);
        const state = renderer.create(screen.getByTestId("main").innerHTML).toJSON();
        expect(state).toMatchSnapshot();
    });


    it("renders the app", () => {
        const app = new App();
        app.componentDidMount = jest.fn(() => {
            Map.render();
        });

        render(<div id="main" data-testid="main">app.render()</div>);
        const element = screen.getByTestId("main");
        expect(element).toBeTruthy();
    });
});

// mock for the fetch data availability method
global.fetch = jest.fn((url) => {
        if(url === "api/available") {
            return Promise.resolve({
                json: () => Promise.resolve(
                        {
                            store_data: "true",
                            predictions_stored: [1654626300, 1654626600, 1654626900, 1654627200, 1654627500, 1654627800,
                                1654628100, 1654628700, 1654629000, 1654629300, 1654629600, 1654629900,
                                1654630200, 1654630500, 1654630800, 1654631100, 1654631400, 1654632000,
                                1654632300, 1654632600, 1654632900, 1654633200, 1654633500, 1654633800,
                                1654634100, 1654634400, 1654635000, 1654635300, 1654635600, 1654635900,
                                1654636200, 1654636500, 1654636800, 1654637100, 1654638300],
                            observations_stored: [1654626300, 1654626600, 1654626900, 1654627200, 1654627500,
                                1654627800, 1654628100, 1654628400, 1654628700, 1654629300, 1654629600, 1654629900,
                                1654630200, 1654630500, 1654630800, 1654631100, 1654631400, 1654631700, 1654632000,
                                1654632300, 1654632600, 1654632900, 1654633200, 1654633500, 1654633800, 1654634100,
                                1654634400, 1654634700, 1654635000, 1654635300, 1654635600, 1654635900, 1654636200,
                                1654636500, 1654636800, 1654637100]
                        }
                    ),
              })
        }
    }
);