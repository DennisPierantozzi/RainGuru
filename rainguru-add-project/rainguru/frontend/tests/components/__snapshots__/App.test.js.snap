// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests matches snapshot 1`] = `"app.render()"`;
